libraries/libfc/include/fc/container/ordered_diff.hpp

```diff
@@ -106,16 +110,24 @@ class ordered_diff {
    requires std::same_as<std::decay_t<X>, diff_result>
    static Container<T> apply_diff(Container<T>&& container, X&& diff) {
       // Remove from the source based on diff.remove_indexes
-      std::ptrdiff_t offset = 0;
-      for (SizeType index : diff.remove_indexes) {
-         FC_ASSERT(index + offset < container.size(), "diff.remove_indexes index ${idx} + offset ${o} not in range ${s}",
-                   ("idx", index)("o", offset)("s", container.size()));
-         container.erase(container.begin() + index + offset);
-         --offset;
+      for (container_size_type i = 0; i < diff.remove_indexes.size(); ++i) {
+         FC_ASSERT(i == 0 || diff.remove_indexes[i] > diff.remove_indexes[i-1],
+                   "diff.remove_indexes not strictly monotonically increasing: current index ${c}, previous index ${p}",
+                   ("c", diff.remove_indexes[i])("p", diff.remove_indexes[i-1]));
+
+         assert(diff.remove_indexes[i] >= i);
+         auto updated_index = diff.remove_indexes[i] - i;
+         FC_ASSERT(updated_index < container.size(), "diff.remove_indexes index ${idx} - i ${i} not in range ${s}",
+                   ("idx", diff.remove_indexes[i])("i", i)("s", container.size()));
+         container.erase(container.begin() + updated_index);
       }
 
       // Insert into the source based on diff.insert_indexes
-      for (auto& [index, value] : diff.insert_indexes) {
+      for (container_size_type i = 0; i < diff.insert_indexes.size(); ++i) {
+         FC_ASSERT(i == 0 || diff.insert_indexes[i].first > diff.insert_indexes[i-1].first,
+                   "diff.insert_indexes not strictly monotonically increasing: current index ${c}, previous index ${p}",
+                   ("c", diff.insert_indexes[i].first)("p", diff.insert_indexes[i-1].first));
+         auto& [index, value] = diff.insert_indexes[i];
          FC_ASSERT(index <= container.size(), "diff.insert_indexes index ${idx} not in range ${s}",
                    ("idx", index)("s", container.size()));
          container.insert(container.begin() + index, std::move(value));
```

Comment:
that `std::move` seems incorrect to me, since we accept `diff` as a forwarding reference.
I think we should `std::forward` `diff` to a local value, and then use that.